Add Release Requirements Policy

t8m opened the pull request:
This adds the policy that covers general requirements that must be met before we create a release (alpha/beta/stable).

It also describes the issue and pull request triage process in short.

t8m commented:
Fixes #22

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,91 @@
+- There is no remaining refactoring required by OMC or OTC for the release.
+- There are no remaining API changes required for the release.
+  _This applies only to beta releases of a major release as API changes
+  are not allowed on minor releases._
```
What about API additions? Arguably that is an API change that is allowed in a minor release - but should not be occurring in a beta release.

t8m replied:
I'll add it as a separate item to make it clear.

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,91 @@
+- There are no remaining API changes required for the release.
+  _This applies only to beta releases of a major release as API changes
+  are not allowed on minor releases._
+- All untriaged issues must be triaged.
```
Untriaged issues that have only just been opened (e.g. in the last 1 or 2 weeks) should perhaps be exempted. Otherwise if someone opens a new issue just before you do the release it would be a blocker.

t8m replied:
I've added `older than 2 days`. IMO the basic triage should be really quick task in most cases so maybe even `older than 1 day` would be OK.

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,91 @@
+  are closed.
+- There is no outstanding untriaged Coverity issue.
+- Coveralls coverage has not decreased overall from the previous release.
+- The CI must pass on tip of the development branch before the release commits
```
The CI must pass on the tip of **the** development branch before the release commits **are** added to the tree.

t8m commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,91 @@
+-----------------
+
+The API and ABI should be stable and the source code should be feature complete
+by the first beta pre-release for the stable release. The following release
```
Sure.

t8m commented:
@mattcaswell fixup pushed addressing your comments.

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- [beta] (pre-)releases
+
+This policy defines the requirements on the state of the source tree that
+must be met before the release can be done.
```
release(s)

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+------------------
+
+As this is just a preview release for testing things that have been worked
+on the development branch of the source code the requirements are minimal.
```
Maybe this instead?

on in the development branch, the requirements are minimal.

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+As this is just a preview release for testing things that have been worked
+on the development branch of the source code the requirements are minimal.
+
+- The CI must pass on tip of the development branch before the release commits
```
on the

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- The tree must be frozen at least for one day, which also ensures the daily CI
+  builds run on the development tree tip.
+- In case of the first beta release the OTC should explicitly approve
+  that the source is ready for a release with a vote.
```
At some point I think performance should be a metric here also.

t8m replied:
When we have ways how to measure performance we can add it. Not now.

kroeckx commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- The code is functionaly complete in regards to the particular release
+  objectives as set by OMC and OTC.
+- There is no remaining refactoring required by OMC or OTC for the release.
+- There are no remaining API changes required for the release.
```
I assume that we can, and do, change the API after the beta, but only in exceptional cases.

t-j-h replied:
If we have planned or expected or anticipated API changes then we have not reached **beta** state. Remember we are operating under the more common definition of beta rather than the previous practice which was effectively later-rolling-alpha releases. 

We can always handle things by an explicit exception vote - but we should not be planning or writing a policy that we expect to never actually follow.

kroeckx replied:
As far as I know, we've never been able to actually do this.

t8m replied:
Anything that is outside of this policy requires exception vote. I think explicitly mention that escape route isn't necessary as that applies to every policy we have.

beldmit replied:
I'd like to see smth like "there are no known/expected/planned remaining API changes"

t8m replied:
Everything is about thinks we know at the point of time we are doing the release. We cannot see into the future.

kroeckx commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+  _Regressions are considered from the previous stable release. The process
+  how OTC assigns the milestone is not exactly defined on purpose._
+- All issues or pull requests with beta milestone for the beta release
+  are closed.
```
It's not really clear what this means to me. Does that mean all bugs must be fixed before the next beta release?

t8m replied:
This supposes that there exists a beta milestone for each beta release. Then when bugs are triaged the individual beta milestones are set. So no, it does not mean all bugs must be fixed, just those with the beta milestone for this beta release that's being released.

beldmit commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- All issues or pull requests with beta milestone for the beta release
+  are closed.
+- There are no outstanding untriaged Coverity issues.
+- Coveralls coverage has not decreased overall from the previous release.
```
Not sure we need this requirement.

t8m replied:
IMO we want it.

beldmit commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+  how OTC assigns the milestone is not exactly defined on purpose._
+- All issues or pull requests with milestone for the release are closed.
+- There are no outstanding untriaged Coverity issues.
+- Coveralls coverage has not decreased overall from the previous release.
```
The same. 
And if still we want to measure it, we also should say that the previous release is the previous stable release.

t8m replied:
Here it is from the previous release on the particular development branch.

paulidale commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- All untriaged issues older than 2 days must be triaged.
+- All regressions on the development branch from which the
+  release is to be done must have a milestone assigned by OTC.
+  _Regressions are considered from the previous stable release. The process
```
The process **of** how the OTC...

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- All untriaged issues older than 2 days must be triaged.
+- All regressions on the development branch from which the
+  release is to be done must have a milestone assigned by OTC.
+  _Regressions are considered from the previous stable release. The process
```
We don't do this at the moment. If we identify a regression in a a stable release, it may or may not be fixed.

Also "regressions are considered from the previous stable relese" - do you mean (say) comparing 3.0 to 1.1.1, or do you mean (say) 3.0.3 compared to 3.02?

t8m replied:
Obviously both should be considered. I'll try to adjust the text.

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+an OTC member according to what is the type (feature, bug, documentation,
+refactoring, ...) of the issue or pull request. When the triage happens for an
+issue or pull request that is a bug/bug fix, it must be assessed whether the
+bug is a regression or not. In general regressions are not allowed in stable
```
We don't do this assessment at the moment - and often it will not be obvious whether it is or not. I don't think we have a way of identifying regressions?

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+- There are no remaining API additions required for the release.
+- All untriaged issues older than 2 days must be triaged.
+- All regressions on the development branch from which the
+  release is to be done must have a milestone assigned by OTC.
```
So what if we identify a regression the day before the release and there is no OTC meeting planned? Does this stop the release?

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+stability requirements as those should be held already by the beta releases.
+
+- All untriaged issues older than 2 days must be triaged.
+- All regressions on the development branch from which the
```
Note that it may not be clear whether something is a regression or not at the time of triage.

t8m replied:
Yeah, triage will never be perfect (as anything else), that does not mean we should not try to triage whether something is a regression.

t8m commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,92 @@
+
+The API and ABI should be stable and the source code should be feature complete
+by the first beta pre-release. The following release requirements apply to beta
+pre-releases.
```
Well... it is not a true release candidate as there might be even known outstanding bugs that we want to fix before the stable release. These requirements do not disallow that.

True release candidate should have no known outstanding bugs that are required to be fixed before the stable release.

slontis replied:
I think @t-j-h made this point during the 3.0 beta.

t8m commented:
Updated to deal with some/most? of the comments from the OTC discussion. Please re-review.

mkoci-koca commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,117 @@
+by the first beta pre-release. The following release requirements apply to beta
+pre-releases.
+
+- The code is functionaly complete in regards to the particular release
```
There is a typo: s/functionaly/functionally

t8m replied:
fixed

t8m commented:
Based on the OTC discussion we had yesterday, I've added a separate Patch releases section. I've also simplified and removed some duplications and moved some things to the Triage process section.

t8m commented:
@openssl/otc please re-review as I plan to start a vote soon.

paulidale commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+
+- [alpha] (pre-)releases
+- [beta] (pre-)releases
+- [major] and [minor] releases
```
These should be broken into two separate lines.
- [major] releases
- [minor] releases

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+  regression or security fixes should be merged during the freeze.
+- For 2 days before the release there should be no changes to ensure the daily
+  CI builds run on the development tree tip.
+- Embargoed security fixes are excepted from the rule above as they cannot
```
Are there more steps associated with security patches (CVE's and notifications)?

t8m replied:
IMO they should not belong here as they would confuse things. If we have any security release process policy, it should be there.

slontis replied:
it still sounds like Release requirements to me..

t8m replied:
I beg to differ.

slontis replied:
In order to do the security release these are requirements - It could be a link to another document...

t8m replied:
We can/should do that later.

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+In general regressions should be fixed as soon as possible, optimally before
+the next release from the development tree is done. However sometimes that
+might not be reasonably possible or an issue is identified to be a regression
+too short time before the release is planned to be done.
```
just before instead of too short time??

t8m replied:
I do not know what people imagine under "just before". To me it sounds like a few hours or something where I meant a few days.

t8m replied:
I've simplified the sentence.

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+fixed before the release is done. That is done by assigning a milestone by
+which the regression must be fixed and the `triaged: OTC evaluated` label.
+
+The triage ideally happens as soon as possible however the triage process can
```
possible,

t8m replied:
Fixed

slontis commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+which the regression must be fixed and the `triaged: OTC evaluated` label.
+
+The triage ideally happens as soon as possible however the triage process can
+be sometimes costly so we aim to have issues triaged no later than 4 days
```
sometimes be costly,

t8m replied:
Fixed

mattcaswell commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,137 @@
+Responsibilities
+----------------
+
+The OTC is responsible to ensure the releases confirm to these requirements.
```
s/confirm/conform/

t8m commented:
Starting the vote for Add Release Requirements Policy at commit [1845646](https://github.com/openssl/technical-policies/pull/40/commits/1845646d25b77ee948e97d0a7fea46cd891d2cba)

mattcaswell commented:
Vote: [+1]

beldmit commented:
Vote: [+1]

slontis commented:
Vote: [+1]

t-j-h commented:
Vote: [+1]

paulidale commented:
Vote: [+1]

levitte commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,138 @@
+- The OTC should explicitly approve that the source is ready for a release with
+  a vote.
+
+Patch Releases
```
It can be noted that we have also call these "release updates", see [policies/stable-release-updates.md](https://github.com/openssl/technical-policies/blob/master/policies/stable-release-updates.md)

levitte replied:
(if we want to change the nomenclature to "patch releases", we should probably do that across the board...  but still leave a note somewhere saying what we used to call it, to connect the dots.  Frankly, I'd welcome such a change, since "patch release" is a broadly used term)

t8m commented:
@levitte @kroeckx ping for vote

levitte commented:
> @levitte @kroeckx ping for vote

As soon as someone answers my question (see review comment)

levitte commented:
Vote: [+1]

mspncp commented:
Vote: [+1]

kroeckx commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,138 @@
+bug is a regression or not.
+
+In general regressions should be fixed as soon as possible, optimally before
+the next release from the development tree is done. However sometimes that
```
It's unclear to me what you mean with development tree. I suggest you make it: "before the next release is done"

t8m replied:
Too late. The vote has started.

kroeckx replied:
You can always create an other PR to fix that.

kroeckx commented on policies/release-requirements.md:
```diff
@@ -0,0 +1,138 @@
+an OTC member according to what is the type (feature, bug, documentation,
+refactoring, ...) of the issue or pull request. When the triage happens for an
+issue or pull request that is a bug/bug fix, it must be assessed whether the
+bug is a regression or not.
```
How do we mark something as regression? Do we just assign it milestones?

t8m replied:
There is now a label: `severity: regression`.

kroeckx replied:
I think this policy should be more clear about such labels.

kroeckx commented:
Voting +1

t8m commented:
Closing the vote.
```
Topic: Add Release Requirements Policy at commit 1845646
       This will become an official OTC policy.
Proposed by: Tomas
Issue link: https://github.com/openssl/technical-policies/pull/40
Public: yes
Opened: 2022-05-25
Closed: 2022-06-06
Accepted: yes   (for: 10, against: 0, abstained: 0, not voted: 0)

   Dmitry     [+1]
   Matt       [+1]
   Pauli      [+1]
   Tim        [+1]
   Richard    [+1]
   Shane      [+1]
   Tomas      [+1]
   Kurt       [+1]
   Matthias   [+1]
   Nicola     [+1]
```
